utils_thai.py

```diff
@@ -193,37 +193,73 @@ def thaipop2(num: float, pos: int) -> str:
+    # on_enname = altnames.merge(provinces, right_index=True,
+    #                            left_on="ProvinceEn").drop(columns=["ProvinceEn_y", "ProvinceEn_x"])
+    # provinces = provinces.combine_first(on_enname)
+
+    # # Add in population data
+    # # popurl = "http://mis.m-society.go.th/tab030104.php?y=2562&p=00&d=0000&xls=y"
+    # popurl = "https://en.wikipedia.org/wiki/Provinces_of_Thailand"
+    # file, _ = next(web_files(popurl, dir="html", check=False))
+    # pop = pd.read_html(file)[2]
+    # pop = pop.join(provinces,
+    #                on="Name(in Thai)").set_index("ProvinceEn").rename(
+    #                    columns={"Population (2019)[1]": "Population"})
+
+    # provinces = provinces.join(pop["Population"], on="ProvinceEn")
+
+    def get_province_mappings_df():
+        def __get_alt_name_mappings(df):
+            """ Return dict of alternative name lookup keys for provinces from the Complete Provinces + Alt Names dataframe/ dataset. 
+                Format: {AltName->Province,..}
+            """
+            alt_names_lookup_dict = df.set_index('Name')[['Alt_names']].to_dict()['Alt_names']
+            r = {}
+            for prov_en,altnames in alt_names_lookup_dict.items():
+                altnames = eval(altnames)
+                if type(altnames) is list and len(altnames) > 0: # Is a list and has entries, therefore add them:
+                    for name in altnames:
+                        if type(name) is str and len(name) > 1: # 
+                            if name not in r:
+                                r[name] = prov_en
+                            elif name in r:
+                                print(f"Warning: duplicate entry of {name} for Province: {prov_en} from Alt Names set: {altnames}")
+                            else:
+                                raise ValueError(f"Unexpected error while iterating over mappings: {name}<-{altnames} for Province: {prov_en}")
+                        else:
+                            raise ValueError(f"Error in alt name: '{name}'. Unexpected error while iterating over mappings: {name}<-{altnames} for Province: {prov_en}")
+            return r
+
+        df = pd.read_csv('province_mapping.csv')
+        map_data = __get_alt_name_mappings(df)
+        map_data = [(k,v) for k,v in map_data.items()]
+        
+        df2 = pd.DataFrame.from_records(map_data, columns=['Alt_names', 'ProvinceEn'])
+        df2 = df2.set_index('ProvinceEn')
+        df3 = df2.join(df.set_index('Name')[['district_num','Name(in Thai)','Population (2019)[1]','Area (km²)[2]']])
+        df3 = df3.reset_index().rename(columns={'index':'ProvinceEn','district_num':'Health District Number',
+                                'Name(in Thai)':'ProvinceTh','Population (2019)[1]':'Population','Area (km²)[2]':'Area_km2'}).set_index('Alt_names')
+        
+        return df3
+
+    provinces = get_province_mappings_df()
```

Comment:
why the extra function?

Reply:
Encapsulation and Isolation of subroutine functionality.

You might have other verification checks or joins to this df within the `get_provinces()` function, but that should not affect the closure code or be polluted by its local variable names.

Reply:
Mitigation of bugs. Readability (good naming). Lowered complexity for reader.... Etc.

Reply:
it litterally one function inside another function with no other code. All its doing is adding an indent.